### PR TITLE
game_list: Add a context menu with "Open Save Location"  option

### PR DESCRIPTION
**Pull request by MerryMage**
What changed:
* Added `Loader::GetProgramId`.
* Added static function `ArchiveSource_SDSaveData::GetSaveDataPathFor`.
* Added signal `GameList::OpenSaveFolderRequested`.

Issues that arose:
* The least messy way I could think of to expose a way to get the save data path was to implement a static function.
* I really really wanted to use a `boost::optional<u64>` to represent the `program_id` since not all formats have a program id.

Screenshot:

### CODE REVIEW

File: src/citra_qt/game_list.cpp [hunk 1]
```diff
@@ -71,6 +73,23 @@ void GameList::DonePopulating() {
...
+    QModelIndex item = tree_view->indexAt(menu_location);
+    if (!item.isValid())
+        return;
+
+    int row = item_model->itemFromIndex(item)->row();
+    QStandardItem* child_file = item_model->invisibleRootItem()->child(row, COLUMN_NAME);
+    u64 program_id = child_file->data(GameListItemPath::ProgramIdRole).toULongLong();
+
+    QMenu context_menu;
+    QAction* open_save_location = context_menu.addAction(tr("Open Save Location"));
```
Comment (bunnei) [inline src/citra_qt/game_list.cpp]: nit: For some reason "Open Save Location" sounds weird to me ...  But perhaps that's just me

File: src/citra_qt/main.cpp [hunk 2]
```diff
@@ -460,6 +464,21 @@ void GMainWindow::OnGameListLoadFile(QString game_path) {
...
+    QString qpath = QString::fromStdString(path);
+
+    QDir dir(qpath);
+    if (!dir.exists()) {
+        QMessageBox::critical(this, tr("Error Opening Save Folder"), tr("Folder does not exist!"));
+        return;
+    }
+
+    LOG_INFO(Frontend, "Opening save data path for program_id=%" PRIu64, program_id);
+    QDesktopServices::openUrl(QUrl::fromLocalFile(qpath));
```
Comment (bunnei) [inline src/citra_qt/main.cpp]: This should work on all three major platforms, right?

File: src/citra_qt/game_list.cpp (hunk 1 above)
Comment (lioncash) [inline src/citra_qt/game_list.cpp]: I've usually seen `Open Folder Location` used.

File: src/core/loader/loader.h [hunk 3]
```diff
@@ -144,6 +144,15 @@ class AppLoader : NonCopyable {
     }
 
     /**
+     * Get the program id of the application
+     * @param out_program_id Reference to store program id into
+     * @return ResultStatus result of function
+     */
+    virtual ResultStatus ReadProgramId(u64& out_program_id) {
```
Comment (bunnei) [inline src/core/loader/loader.h]: So why didn't you use boost::optional? (or std::optional, is that not available on all compilers yet?)
I see you're just following the pattern already used in this file, so I guess it's fine for now, but the loader interface could use some improvements

File: src/citra_qt/main.cpp (hunk 2 above)
Comment (MerryMage) [inline src/citra_qt/main.cpp]: Yes.

File: src/core/loader/loader.h (hunk 3 above)
Comment (MerryMage) [inline src/core/loader/loader.h]: `std::optional` is part of C++17.

I imitated the pattern already in the file. I could do a follow up PR that improves on the interface.

File: src/citra_qt/game_list.cpp (hunk 1 above)
Comment (MerryMage) [inline src/citra_qt/game_list.cpp]: Reworded to `Open Save Data Location`.

**Comment by MerryMage** [comment]
Should update [the guide on the wiki](https://github.com/citra-emu/citra/wiki/Dumping-Save-Data-from-a-3DS-Console) if/when this is merged.

**Comment by MerryMage** [comment]
Should the folder be created if it doesn't exist?

**Comment by legoj15** [comment]
I think that would be a good idea

**Comment by yuriks** [comment]
I think directory creation should be in a separate PR. (I'm not sure if it should even *be* done, since that'll create problems with the lack of save metadata. @Subv says many things actually rely on that being correct.)

**Comment by yuriks** [comment]
@MerryMage feel free to merge this if you're done with it

**Pull request merged**